### DISCUSSION TITLE
Can unikraft core directly interact with Hardware with only a hypervisor?

### DISCUSSION THREAD

**Discussion by chenzq-98** [unikraft/unikraft#620] [Q&A]
In session2 https://unikraft.org/community/hackathons/sessions/behind-scenes/. 

Figure of Unikraft over Type 1 hypervisor says that unikraft can directly interact hardware with only a hypervisor. However, in the examples shown in sessions, we run the images with qemu-kvm, which is not the case of Type 1. How to execute our images with the Type 1 architecture?

or more detailed, can unikraft run on bare metal?

**Comment by nderjung** [comment]
Hi @czq693497091,

Unikraft can run as a VM on either Type-I or Type-II hypervisors as well as bare metal thanks to the "platform" microlibrary architecture.  At present, QEMU/KVM is the most stable and best supported "platform" offered by Unikraft.  KVM is a Type-II hypervisor.  The next best supported platform is Xen, which is a Type-I hypervisor.  It is also possible to run Unikraft directly on bare metal without a hypervisor, e.g. on a Raspberry Pi.  

To build Unikraft for Xen, simply open up the `make menuconfig` and select the Xen platform.  This way you will be able to target a Type-I hypervisor.

**Reply by Daparrag** [reply]
@nderjung, please correct me if I'm mistaken, but in order for Unikraft to interact with physical hardware, it must run as a bare-metal system, as it has direct access to physical resources. In all other cases, hypervisor support is required, and there are three possible options:

1. The hypervisor can emulate physical devices and provide an interface to the Unikraft-VM.

2. The hypervisor can use virtio as a medium for interaction between the Unikraft-VM and the physical device.

3. Devices can be passed directly to the Unikraft-VM, as Xen does with passthrough.

**Reply by nderjung** [reply]
Type-I hypervisors, such as Xen, essentially partition hardware resources and allow the guest to access the resources directly. This is one way to have the unikernel "touch" the hardware. Of course, this requires then therefore a type-I hypervisor to be present on the machine beforehand. Using Xen, for instance, is fine for many devices, because you can essentially have a thin controller that then segments the rest of the hardware resources for guests. However, devices that do not have much physical resources, such as IoT devices, having this controller domain is not possible and Unikraft can (though it has limited support) run as the only entity. This also means that Unikraft would be "touching" the hardware. However, the drawbacks here are that you essentially have to continuously flash the ROM or boot device to update the kernel/Unikraft.

That said, to answer your points: 1. Yes, 2. Yes. and 3. Yes. :) Almost all levels of abstraction are possible. Xen offers the closest ability to speak with the hardware, introducing "hypercalls" which are loaded into registers which trap code that would otherwise perform "out of bounds" calls from a segmented-hardware point-of-view.

**Comment by chenzq-98** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        In the session, I don’t see the demo of start a unikraftcore with Xen. I choose Xen in platform and generate app-helloworld_xen-x86_64. How to start this vm?
    
        
            
        
    
    
      
    
        
                    
            
                
                    
                            
                                
                            
                            
                                693497091
                            
                    
                        
                            
                                    ***@***.***
                            
                        
                
            
        
        
    
 
    

  
    ---- Replied Message ----
  
  
    
      
        
         From 
        
        
            Alexander ***@***.***>
            
        
      
      
        
         Date 
        
        
        11/2/2022 01:26
        
      
      
        
         To 
        
        
         
          
            ***@***.***>
            
          
        
      
      
        
         Cc 
        
        
          
            Zhenqian ***@***.***>
            ,
          
          
            ***@***.***>
            
          
        
      
      
        
         Subject 
        
        
              Re: [unikraft/unikraft] Can unikraft core directly interact with Hardware with only a hypervisor? (Discussion #620)
        
      
    
  
  
Hi @czq693497091,
Unikraft can run as a VM on either Type-I or Type-II hypervisors as well as bare metal thanks to the "platform" microlibrary architecture.  At present, QEMU/KVM is the most stable and best supported "platform" offered by Unikraft.  KVM is a Type-II hypervisor.  The next best supported platform is Xen, which is a Type-I hypervisor.  It is also possible to run Unikraft directly on bare metal without a hypervisor, e.g. on a Raspberry Pi.
To build Unikraft for Xen, simply open up the make menuconfig and select the Xen platform.  This way you will be able to target a Type-I hypervisor.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by razvand** [comment]
Everyone, since we are already (viciously) using [the Unikraft Discord server](https://bit.ly/UnikraftDiscord) for discussions, meetings, brainstorming and jokes (mostly bad, we do what we can), we will close the Discussions tab. We will do this on Friday, March 31, 2023.

Please move any discussions or questions on [the Unikraft Discord server](https://bit.ly/UnikraftDiscord). See you all there! :handshake:
